aptos-move/aptos-aggregator/src/transaction.rs

```diff
@@ -22,9 +23,24 @@ impl ChangeSetExt {
         }
     }
 
+    pub fn delta_change_set(&self) -> &DeltaChangeSet {
+        &self.delta_change_set
+    }
+
+    pub fn write_set(&self) -> &WriteSet {
+        self.change_set.write_set()
+    }
+
     pub fn into_inner(self) -> (DeltaChangeSet, ChangeSet) {
         (self.delta_change_set, self.change_set)
     }
+
+    pub fn squash(self, other: Self) -> anyhow::Result<Self> {
```

Comment:
Where are we covering that case when delta change set of one has an intersection with the change set of the other?

Reply:
you want something like: move_vm_ext/session.rs logic from line 273, I think (@georgemitenkov can confirm)

Reply:
But I need to merge two `DeltaChangeSet`, not `AggregatorChangeSet`

Reply:
True, but change set contains full writes and delta change set contains deltas.
Both ChangeSetExts each contain a change set container<(key, Write)> and delta change set container<(key, Delta>).
You want an output ChangeSetExt.

For any given key, if it appears in only one you take it. If it appears in both, there has to be a merging logic:
delta, write -> write.
write, delta -> apply delta, unless write was a delete, which probably should be unreachable.
write, write -> second write, squashing writes should do this.
delta, delta -> merge_with of the deltas, squashing deltas should do this.

Does this make sense?

Reply:
If you are saying that I should merge the two `ChangeSet`s using the logic above, then I fully agree and that's exactly what I have implemented.

Wait are you suggesting that the `ChangeSet` and `DeltaChangeSet` are somehow connected???

Reply:
Oh so this is what you mean by "that case when delta change set of one has an intersection with the change set of the other"

Reply:
for example, if sets are
1: (0, +100), (1, write 5)
1: (0, write 3), (1, +100)

the output is (0, write 3), (1, write 105)

Reply:
Within one change set, we guarantee that deltas and writes have different keys, but when merging 2 there can be intersection as @gelash said

Reply:
"we guarantee that deltas and writes have different keys" - because if aggregator knows a value - it produces a write, so this transaction cannot possibly produce a delta for the same state key associated with this aggregator

Reply:
So let's say there are two `ChangeSetExt`s `ext1` and `ext2`. Is it correct that:
1. The `ChangeSet` and `DeltaChangeSet` in each `ChangeSetExt` have disjoint keys? This seems to be the case according to the code in `SessionOutput::into_change_set`.
2. `ext1.change_set` could overlap with `ext2.change_set` or `ext2.delta_change_set`, but not both; `ext1.delta_change_set` could overlap with ext2.change_set` or `ext2.delta_change_set`, but not both?

Reply:
I see. I think I understand the problem. Let me try to implement the correct merger algorithm.

Reply:
1. Yes
2. Yes

Reply:
I thought we squash the aggregator changeset first, the logic seems be there